opm/autodiff/MSWellHelpers.hpp

```diff
@@ -0,0 +1,195 @@
+
+namespace mswellhelpers
+{
+
+    // obtain y = D^-1 * x with a direct solver
+    template <typename MatrixType, typename VectorType>
+    VectorType
+    invDXDirect(const MatrixType& D, VectorType x)
+    {
+        VectorType y(x.size());
+        y = 0.;
+
+        Dune::UMFPack<MatrixType> linsolver(D, 0);
+
+        // Object storing some statistics about the solving process
+        Dune::InverseOperatorResult res;
+
+        // Solve
+        linsolver.apply(y, x, res);
+
+        // Checking if there is any inf or nan in y
+        // it will be the solution before we find a way to catch the singularity of the matrix
+        for (int i_block = 0; i_block < y.size(); ++i_block) {
+            for (int i_elem = 0; i_elem < y[i_block].size(); ++i_elem) {
+                if (std::isinf(y[i_block][i_elem]) || std::isnan(y[i_block][i_elem]) ) {
+                    OPM_THROW(Opm::NumericalProblem, "nan or inf value found in invDXDirect due to singular matrix");
+                }
+            }
+        }
+
+        return y;
+    }
+
+
+
+
+
+    // obtain y = D^-1 * x with a BICSSTAB iterative solver
+    template <typename MatrixType, typename VectorType>
+    VectorType
+    invDX(const MatrixType& D, VectorType x)
+    {
+        // the function will change the value of x, so we should not use reference of x here.
+
+        // TODO: store some of the following information to avoid to call it again and again for
+        // efficiency improvement.
+        // Bassically, only the solve / apply step is different.
+
+        VectorType y(x.size());
+        y = 0.;
+
+        Dune::MatrixAdapter<MatrixType, VectorType, VectorType> linearOperator(D);
+
+        // Sequential incomplete LU decomposition as the preconditioner
+        Dune::SeqILU0<MatrixType, VectorType, VectorType> preconditioner(D, 1.0);
+        // Dune::SeqILUn<MatrixType, VectorType, VectorType> preconditioner(D, 1, 0.92);
+        // Dune::SeqGS<MatrixType, VectorType, VectorType> preconditioner(D, 1, 1);
+        // Dune::SeqJac<MatrixType, VectorType, VectorType> preconditioner(D, 1, 1);
+
+        // Preconditioned BICGSTAB solver
+        Dune::BiCGSTABSolver<VectorType> linsolver(linearOperator,
+                                                   preconditioner,
+                                                   1.e-8, // desired residual reduction factor
+                                                   250, // maximum number of iterations
+                                                   0); // verbosity of the solver */
+
+        // Object storing some statistics about the solving process
+        Dune::InverseOperatorResult res;
+
+        // Solve
+        linsolver.apply(y, x, res);
+
+        if ( !res.converged ) {
+            OPM_THROW(Opm::NumericalProblem, "the invDX does not get converged! ");
+        }
+
+        return y;
+    }
+
+
+
+
+
+    static double haalandFormular(const double re, const double diameter, const double roughness)
```

Comment:
Instead of being `static`, these functions (non-template functions that are defined inline in a header) should be declared `inline`, which will give them weak linkage.